### PR TITLE
Add address derivation

### PR DESCRIPTION
**Pull request by paweljakubas**
Add address derivation

- [x] Add abstraction

### CODE REVIEW

File: cardano-addresses.cabal [hunk 1]
```diff
@@ -40,6 +40,7 @@ library
     ghc-options:
       -Werror
   exposed-modules:
+      AddressDerivation
```
Comment (KtorZ) [inline cardano-addresses.cabal:43]: What about `Cardano.KeyDerivation` :) ?

File: src/AddressDerivation.hs [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+import Data.Word
+    ( Word32 )
+import Fmt
+    ( Buildable (..) )
+import GHC.Generics
+    ( Generic )
+
+-- $overview
+--
+-- These abstractions when combined together enable address derivation
```
Comment (KtorZ) [inline src/AddressDerivation.hs:59]: That's quite a quick overview :smile:

File: src/AddressDerivation.hs [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+        -> AccountingStyle
+        -> Index 'Soft 'AddressK
+        -> key 'AddressK XPub
+
+
+-- | Abstract interface for constructing a /Master Key/.
+class GenMasterKey (key :: Depth -> * -> *) where
+
+    -- | Generate a root key from a corresponding seed.
+    generateKeyFromSeed :: SomeMnemonic -> ScrubbedBytes -> key 'RootK XPrv
```
Comment (KtorZ) [inline src/AddressDerivation.hs:215]: Note that Icarus' and Ledger' style allows for defining an extra second factor passphrase. We could make this a type family as:

```hs
class GenMasterKey (key :: Depth -> * -> *) where
    type GenMasterKeyFrom key :: *

    -- | Generate a root key from a corresponding seed.
    genMasterKey :: GenMasterKeyFrom key -> ScrubbedBytes -> key 'RootK XPrv
```

File: stack.yaml [hunk 4]
```diff
@@ -1,8 +1,8 @@
-resolver: lts-14.25
+resolver: https://raw.githubusercontent.com/input-output-hk/cardano-haskell/master/snapshots/cardano-1.9.3.yaml
```
Comment (KtorZ) [inline stack.yaml:1]: :+1:

File: src/AddressDerivation.hs [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    -- derivation scheme 2 (see <https://github.com/input-output-hk/cardano-crypto/ cardano-crypto>
+    -- package for more details).
+    --
+    -- It is preferred to use 'deriveAddressPublicKey' whenever possible to avoid
+    -- having to manipulate passphrases and private keys.
+    --
+    -- NOTE: The caller is expected to provide the corresponding passphrase (and
+    -- to have checked that the passphrase is valid). Providing a wrong passphrase
+    -- will not make the function fail but will instead, yield an incorrect new
+    -- key that doesn't belong to the wallet.
```
Comment (KtorZ) [inline src/AddressDerivation.hs:189]: We might want to change this in this library ^^"...

File: cardano-addresses.cabal (hunk 1 above)
Comment (paweljakubas) [reply cardano-addresses.cabal:43]: changed

File: src/AddressDerivation.hs (hunk 2 above)
Comment (paweljakubas) [reply src/AddressDerivation.hs:59]: in this PR was focusing on the code. hope to spend more time on docs in the next PRs

File: src/AddressDerivation.hs (hunk 3 above)
Comment (paweljakubas) [reply src/AddressDerivation.hs:215]: indeed, as thinking about Byron

File: src/AddressDerivation.hs (hunk 5 above)
Comment (paweljakubas) [reply src/AddressDerivation.hs:189]: removed comments

**Pull request merged**